### PR TITLE
Refactor PublicTask into a decorator task

### LINKED ISSUE #3974: Remove possibility for shared state in PublicTasks and friends
Similar to the changes in #3946, we should break out the task steps to a
separate class to encapsulate state. Tests first will help determine if this is
prone to the same problem as the project.task Task subclasses.

### PR DESCRIPTION
**Pull request by stsewd**
~Still wip, I need to update some docstrings~, but I already tested it, and it works :)

Fix #3974 and ref to #3973

### CODE REVIEW

File: readthedocs/oauth/tasks.py [hunk 1]
```diff
@@ -22,20 +31,12 @@
...
-                service.sync()
-
-
-sync_remote_repositories = SyncRemoteRepositories()
+@app.task(queue='web', base=PublicTask)
+def sync_remote_repositories(user_id):
+    user = User.objects.get(pk=user_id)
+    for service_cls in registry:
+        for service in service_cls.for_user(user):
+            service.sync()
```
Comment (stsewd) [inline readthedocs/oauth/tasks.py]: The api for defining a public task is cleaner now :D

File: readthedocs/oauth/tasks.py [hunk 2]
```diff
@@ -22,20 +31,12 @@
 
 
 @permission_check(user_id_matches)
```
Comment (stsewd) [inline readthedocs/oauth/tasks.py]: In the celery docs, it says that we shouldn't do this (put the task decorator down), but the task decorator overrides all function's attributes and deletes the added by `permission_check`, this way we can keep the added attribute.

**Comment by stsewd** [comment]
I believe this is ready, I also tested it locally, it works as expected.

File: readthedocs/oauth/tasks.py (hunk 1 above)
Comment (agjohnson) [reply readthedocs/oauth/tasks.py]: Much nicer!

**Pull request merged**